mixnet/client/src/receiver.rs

```diff
@@ -15,40 +15,93 @@ pub struct Receiver;
-    ) -> Result<(), Box<dyn Error>> {
-        let mut socket = TcpStream::connect(node_address).await?;
-        let mut message_reconstructor: MessageReconstructor = Default::default();
+    ) -> Result<impl Stream<Item = Result<Vec<u8>, Box<dyn Error>>> + Send + 'static, Box<dyn Error>>
+    {
+        let socket = TcpStream::connect(node_address).await?;
+
+        Ok(Self::message_stream(Box::pin(Self::fragment_stream(
+            socket,
+        ))))
+    }
+
+    fn fragment_stream(
+        socket: TcpStream,
+    ) -> impl Stream<Item = Result<Fragment, Box<dyn Error>>> + Send + 'static {
+        stream::unfold(socket, |mut socket| async move {
+            let Ok(body) = Body::read(&mut socket).await else {
+                // TODO: Maybe this is a hard error and the stream is corrupted? In that case stop the stream
+                return Some((Err("Could not read body from socket".into()), socket));
+            };
 
-        loop {
-            let body = Body::read(&mut socket).await?;
             match body {
-                Body::SphinxPacket(_) => return Err("received sphinx packet not expected".into()),
-                Body::FinalPayload(payload) => {
-                    Self::handle_payload(payload, message_tx.clone(), &mut message_reconstructor)
-                        .await?
+                Body::SphinxPacket(_) => {
+                    Some((Err("received sphinx packet not expected".into()), socket))
                 }
+                Body::FinalPayload(payload) => Some((Self::fragment_from_payload(payload), socket)),
             }
-        }
+        })
+    }
+
+    fn message_stream(
+        fragment_stream: impl Stream<Item = Result<Fragment, Box<dyn Error>>> + Send + Unpin + 'static,
+    ) -> impl Stream<Item = Result<Vec<u8>, Box<dyn Error>>> + Send + 'static {
+        // MessageReconstructor buffers all received fragments
+        // and eventually returns reconstructed messages.
+        let message_reconstructor: MessageReconstructor = Default::default();
+
+        stream::unfold(
+            (fragment_stream, message_reconstructor),
+            |(mut fragment_stream, mut message_reconstructor)| async move {
+                let result =
+                    Self::reconstruct_message(&mut fragment_stream, &mut message_reconstructor)
+                        .await;
+                Some((result, (fragment_stream, message_reconstructor)))
+            },
+        )
+    }
+
+    fn fragment_from_payload(payload: Payload) -> Result<Fragment, Box<dyn Error>> {
+        Ok(Fragment::try_from_bytes(&payload.recover_plaintext()?)?)
     }
 
-    async fn handle_payload(
-        payload: Payload,
-        mut message_tx: impl Sink<Vec<u8>> + Unpin,
+    async fn reconstruct_message(
+        fragment_stream: &mut (impl Stream<Item = Result<Fragment, Box<dyn Error>>>
+                  + Send
+                  + Unpin
+                  + 'static),
```

Comment:
why the `&mut`? if this consumes the other stream it may as well take ownership of it.

Reply:
It was because this `reconstruct_message` function is called multiple times by `stream::unfold`, and the `fragment_stream` is owned by the `stream::unfold` (as a state). 

I may be able to make this `reconstruct_message` function take the ownership of `fragment_stream` and return it again, but I thought that taking reference is simpler. What do you think about this?

Reply:
To be honest, I'm also okay with removing `reconstruct_message` function and moving its logic into the `stream::unfold`. That would be cleaner.

Reply:
But you can move the stream to consume it, then use a `&mut` inside were you exactly need it 🤔

Reply:
I think I implemented it as you suggested. Let me check it again.

Reply:
No problem. Thank you for taking a look at details!